numismatic/feeds/base.py

```diff
@@ -245,52 +249,33 @@ class WebsocketClient(abc.ABC):
 
     exchange = None
     websocket_url = None
-    _websocket_map = dict()
-    _semaphore_map = dict()
-
-    @property
-    def websocket(self):
-        try:
-            return self._websocket_map[self.websocket_url]
-        except KeyError:
-            return None
-    
-    @property
-    def semaphore(self):
-        if self.websocket_url is None:
-            return None
-        try:
-            return WebsocketClient._semaphore_map[self.websocket_url]
-        except KeyError:
-            semaphore = asyncio.BoundedSemaphore(1)
-            WebsocketClient._semaphore_map[self.websocket_url] = semaphore
-            return semaphore
-        except Exception as ex:
-            logger.error(ex)
-            logger.error(packet)
-            raise
-
-    async def _connect(self):
+    _websocket_future = None
+    __listener = None
+    subscriptions = []
+            
+    def __attrs_post_init__(self):
+        asyncio.ensure_future(self._listener())
+
+    async def _connect(self, websocket_url=None):
         '''
-            Connects to websocket. Uses a semaphore
-            to ensure that only one connection at
-            a time will happen
+            Connects to websocket. Uses a future to ensure that only one
+            connection at a time will happen
         '''
-        with (await self.semaphore):
-            websocket = self.websocket
-            if websocket is not None and\
-               websocket.open:
-                return
-
+        if websocket_url is None:
             websocket_url = self.websocket_url
+        if self._websocket_future is None:
             logger.info(f'Connecting to {websocket_url!r} ...')
-            websocket = await websockets.connect(websocket_url)
-            self._websocket_map[websocket_url] = websocket
-            if hasattr(self, 'on_connect'):
-                await self.on_connect(WebsocketClient.websocket)
```

Comment:
I think `on_connect` was only being used by Bitfinex and that use case could be handled by the existing message handler mechanism so I removed this special case.